src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,475 @@
+            all_samples, all_params = self._obtain_samples(runner)
+            tabular.record('TotalEnvSteps', runner.total_env_steps)
+            last_return = self.train_once(runner, all_samples, all_params)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, all_samples, all_params):
+        """Train the algorithm once.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): A two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            all_params (list[dict]): A list of named parameter dictionaries.
+                Each dictionary contains key value pair of names (str) and
+                parameters (torch.Tensor).
+
+        Returns:
+            float: Average return.
+
+        """
+        old_theta = dict(self.policy.named_parameters())
+
+        kl_before = self._compute_kl_constraint(itr,
+                                                all_samples,
+                                                all_params,
+                                                set_grad=False)
+
+        meta_objective = self._compute_meta_loss(itr, all_samples, all_params)
+
+        self._meta_optimizer.zero_grad()
+        meta_objective.backward()
+
+        self._meta_optimize(itr, all_samples, all_params)
+
+        # Log
+        loss_after = self._compute_meta_loss(itr,
+                                             all_samples,
+                                             all_params,
+                                             set_grad=False)
+        kl_after = self._compute_kl_constraint(itr,
+                                               all_samples,
+                                               all_params,
+                                               set_grad=False)
+
+        with torch.no_grad():
+            policy_entropy = self._compute_policy_entropy(
+                [task_samples[0] for task_samples in all_samples])
+            average_return = self.evaluate_performance(
+                itr, all_samples, meta_objective.item(), loss_after.item(),
+                kl_before.item(), kl_after.item(),
+                policy_entropy.mean().item())
+
+        update_module_params(self._old_policy, old_theta)
+
+        return average_return
+
+    def _obtain_samples(self, runner):
+        """Obtain samples for each task before and after the fast-adaptation.
+
+        Args:
+            runner (LocalRunner): A local runner instance to obtain samples.
+
+        Returns:
+            tuple: Tuple of (all_samples, all_params).
+                all_samples (list[MAMLTrajectoryBatch]): A list of size
+                    [meta_batch_size * (num_grad_updates + 1)]
+                all_params (list[dict]): A list of named parameter
+                    dictionaries.
+
+        """
+        tasks = self._env.sample_tasks(self._meta_batch_size)
```

Comment:
so here you pass in a single environment and set task when needed, while another way is to pass multiple environments at the first place. Does that matter? I'm now doing the second approach for RL^2 -- should we keep it consistent and if so, which one is better? @ryanjulian @krzentner

Reply:
please use an API which assumes each task is an independent `gym.Env` object.

Reply:
I am still not sure where should I get tasks. Should I expect sample_tasks() to be called outside of MAML? Or just leave the current implementation here and wait for KR's meta-task sampler.

Reply:
perhaps, for now, you can subclass runner so that it calls `sample_task/set_task` for you after every itr.

@krzentner PTAL

Reply:
I think in this case it will just be easy to use the new sampler. The performance might be slightly worse using `LocalSampler`, but there should be any code changes required.